File: mobile-test-agent/mobile_test_agent/agent.py
"""Orchestrator: the find → act → verify loop.

Escalation ladder per step (cheapest first — this is the token budget):

  parse:   grammar (0 tokens) ────────────► LLM parse (~150 tokens, rare)
  find:    fuzzy matcher (0 tokens) ─► scroll-to-find (0) ─► LLM pick
           (~300 tokens, rare) ─► appium_ai vision (optional)
  act:     deterministic executors (0 tokens)
  verify:  postcondition on fresh snapshot (0 tokens)
  recover: wait-for-idle → re-find → alternate action path → retry (×N)

A typical flow therefore costs *zero* LLM tokens; the LLM only pays for
genuinely ambiguous screens, which is how this stays far below
plan-every-step agents (DroidRun averages ~3,225 tokens per task).
"""
from __future__ import annotations

import asyncio
import json
import logging
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import List, Optional

from .actions import (
    ActionError, Device, scroll_to_find, select_option, set_date, set_toggle,
)
from .config import Config
from .llm import (
    COMMAND_SCHEMA, DISAMBIG_SCHEMA, LLMUnavailable, OllamaLLM,
)
from .matcher import MatchResult, resolve
from .mcp_client import AppiumMCP, ToolCallError
from .parser import Command, parse_date_value, parse_step
from .snapshot import Snapshot, UIElement
from . import verify as V

log = logging.getLogger(__name__)


@dataclass
class StepResult:
    step: str
    status: str            # PASSED | PASSED_UNVERIFIED | FAILED | SKIPPED
    attempts: int = 0
    detail: str = ""
    duration_s: float = 0.0
    llm_used: bool = False


@dataclass
class RunReport:
    results: List[StepResult] = field(default_factory=list)
    token_usage: dict = field(default_factory=dict)

    @property
    def passed(self) -> bool:
        return all(r.status.startswith("PASSED") or r.status == "SKIPPED"
                   for r in self.results)

    def as_dict(self) -> dict:
        return {
            "passed": self.passed,
            "steps": [r.__dict__ for r in self.results],
            "token_usage": self.token_usage,
        }


class MobileTestAgent:
    def __init__(self, cfg: Config) -> None:
        self.cfg = cfg
        self.mcp = AppiumMCP(cfg)
        self.llm = OllamaLLM(cfg)
        self.dev = Device(self.mcp, cfg)

    # ---- lifecycle ------------------------------------------------------
    async def start(self) -> None:
        await self.mcp.start()
        if self.cfg.create_session:
            await self._ensure_session()

    async def _ensure_session(self) -> None:
        caps = self.cfg.load_capabilities()
        try:
            await self.mcp.try_call("session", [
                {"action": "create", "capabilities": caps},
                {"action": "create"},
                {"capabilities": caps},
            ])
        except ToolCallError as exc:
            log.warning("session create failed (%s); assuming server "
                        "auto-manages the session", exc)

    async def close(self) -> None:
        await self.mcp.close()

    # ---- flow running -----------------------------------------------------
    async def run_flow(self, steps: List[str]) -> RunReport:
        report = RunReport()
        for step in steps:
            step = step.strip()
            if not step or step.startswith("#"):
                continue
            result = await self.run_step(step)
            report.results.append(result)
            log.info("%-18s %s  (%s)", result.status, step, result.detail)
            if result.status == "FAILED" and not self.cfg.continue_on_failure:
                break
        report.token_usage = self.llm.usage.as_dict()
        return report

    async def run_step(self, text: str) -> StepResult:
        t0 = time.monotonic()
        result = StepResult(step=text, status="FAILED")

        cmd = parse_step(text, self.cfg.day_first_dates)
        if cmd is None:
            cmd = self._llm_parse(text)
            result.llm_used = cmd is not None
        if cmd is None:
            result.detail = "could not understand step (grammar + LLM failed)"
            result.duration_s = time.monotonic() - t0
            return result

        deadline = t0 + self.cfg.step_timeout_s
        last_detail = ""
        for attempt in range(1, self.cfg.max_attempts + 1):
            result.attempts = attempt
            try:
                verification = await self._attempt(cmd, result)
                if verification.ok:
                    result.status = "PASSED_UNVERIFIED" if verification.weak else "PASSED"
                    result.detail = verification.detail
                    result.duration_s = time.monotonic() - t0
                    return result
                last_detail = verification.detail
            except (ActionError, ToolCallError) as exc:
                last_detail = str(exc)
            if time.monotonic() > deadline:
                last_detail += " (step timeout)"
                break
            # recovery between attempts: settle UI, drop keyboard, re-snapshot
            await self.dev.hide_keyboard()
            await self.dev.stable_snapshot(timeout=3)
            log.info("retrying step (attempt %d): %s — %s", attempt + 1, text, last_detail)

        result.detail = last_detail
        result.duration_s = time.monotonic() - t0
        return result

    # ---- one find→act→verify attempt --------------------------------------
    async def _attempt(self, cmd: Command, result: StepResult) -> V.Verification:
        dev, cfg = self.dev, self.cfg

        # commands with no target element
        if cmd.kind == "wait":
            await asyncio.sleep(float(cmd.value or 1))
            return V.Verification(True, "waited")
        if cmd.kind == "back":
            await self.mcp.try_call("device", [
                {"action": "back"}, {"action": "press_back"}, {"key": "back"}])
            return V.Verification(True, "back pressed", weak=True)
        if cmd.kind == "home":
            await self.mcp.try_call("device", [{"action": "home"}, {"key": "home"}])
            return V.Verification(True, "home pressed", weak=True)
        if cmd.kind == "hide_keyboard":
            await dev.hide_keyboard()
            return V.Verification(True, "keyboard hidden", weak=True)
        if cmd.kind == "launch":
            await self.mcp.try_call("app_lifecycle", [
                {"action": "launch", "appId": cmd.target},
                {"action": "activate", "appId": cmd.target},
                {"appId": cmd.target},
            ])
            await dev.stable_snapshot()
            return V.Verification(True, f"launched {cmd.target}", weak=True)
        if cmd.kind == "press_key":
            await self.mcp.try_call("keyboard", [
                {"action": "press", "key": cmd.value}, {"key": cmd.value}])
            return V.Verification(True, f"pressed {cmd.value}", weak=True)

        # assertions: wait-for-condition rather than instant check
        if cmd.kind in ("assert_visible", "assert_not_visible"):
            deadline = time.monotonic() + cfg.wait_for_element_timeout_s
            while True:
                snap = await dev.snapshot()
                ver = V.verify_visible(snap, cmd, cfg.match_accept_score)
                if ver.ok or time.monotonic() > deadline:
                    return ver
                await asyncio.sleep(cfg.stable_poll_interval_s)

        # bare directional scroll
        if cmd.kind == "scroll" and not cmd.target:
            snap = await dev.stable_snapshot(timeout=3)
            containers = snap.scroll_containers()
            await dev.swipe_dir(cmd.direction or "down",
                                containers[0] if containers else None, snap)
            return V.Verification(True, f"scrolled {cmd.direction}")

        # ---------- FIND ----------
        before = await dev.stable_snapshot()
        match = resolve(cmd, before, cfg.match_accept_score,
                        cfg.match_accept_margin, cfg.match_strong_score)
        if not match.confident and cmd.kind != "select":
            match, before = await scroll_to_find(dev, cmd, cfg)
        if not match.confident and match.candidates:
            picked = self._llm_disambiguate(cmd, before, match)
            if picked is not None:
                match = MatchResult(picked, 1.0, 1.0, match.candidates)
                result.llm_used = True
        if not match.confident:
            if cmd.kind == "select" and cmd.target is None:
                pass  # select can work from the option value alone
            elif cfg.enable_vision_fallback:
                return await self._vision_fallback(cmd, result)
            else:
                return V.Verification(
                    False,
                    f"element not found for '{cmd.target}' "
                    f"(best score {match.score:.2f})",
                )
        el = match.element

        # ---------- ACT + VERIFY ----------
        if cmd.kind in ("tap", "long_press"):
            if cmd.kind == "tap":
                await dev.tap(el)
            else:
                await dev.long_press(el)
            after = await dev.stable_snapshot()
            return V.verify_tap(before, after, el)

        if cmd.kind == "input" and not cmd.maybe_date:
            await dev.type_text(el, cmd.value or "")
            after = await dev.stable_snapshot(timeout=4)
            ver = V.verify_input(after, el, cmd)
            if ver.ok:
                await dev.hide_keyboard()
            return ver

        if cmd.kind in ("set_date",) or (cmd.kind == "input" and cmd.maybe_date):
            ok = await set_date(dev, cmd, cfg, before, el)
            if not ok:
                return V.Verification(False, "date picker interaction failed")
            after = await dev.stable_snapshot(timeout=4)
            d = parse_date_value(cmd.value or "", cfg.day_first_dates)
            variants = []
            if d:
                variants = [d.strftime(f) for f in
                            ("%d/%m/%Y", "%m/%d/%Y", "%d %b %Y", "%B %d, %Y",
                             "%Y-%m-%d", "%d-%m-%Y")]
            return V.verify_date_shown(after, el, variants)

        if cmd.kind == "toggle":
            ok = await set_toggle(dev, el, cmd.toggle_state)
            return V.Verification(ok, "toggle state verified" if ok
                                  else "toggle did not reach desired state")

        if cmd.kind == "select":
            ok = await select_option(dev, cmd, cfg, before, el)
            if not ok:
                return V.Verification(False, f"option '{cmd.value}' not found")
            after = await dev.stable_snapshot(timeout=4)
            shown = V.verify_visible(
                after, Command(kind="assert_visible", target=cmd.value, raw=cmd.raw),
                cfg.match_accept_score)
            return V.Verification(True, f"selected '{cmd.value}'",
                                  weak=not shown.ok)

        if cmd.kind == "scroll":  # scroll to <target> — finding it IS the pass
            return V.Verification(True, f"scrolled to '{cmd.target}'")

        return V.Verification(False, f"unsupported command kind: {cmd.kind}")

    # ---- LLM escalations ------------------------------------------------
    def _llm_parse(self, text: str) -> Optional[Command]:
        try:
            data = self.llm.chat_json(
                system=(
                    "You compile one mobile-UI test step into one JSON command. "
                    "kinds: tap, long_press, input, set_date, select, toggle, "
                    "scroll, assert_visible, assert_not_visible, wait, back, "
                    "home, hide_keyboard, launch, press_key. "
                    "target = element description; value = text/option/date."
                ),
                user=text,
                schema=COMMAND_SCHEMA,
            )
        except LLMUnavailable as exc:
            log.warning("LLM parse unavailable: %s", exc)
            return None
        if not data.get("kind"):
            return None
        return Command(
            kind=data["kind"], target=data.get("target"),
            value=data.get("value"), direction=data.get("direction"),
            toggle_state=data.get("toggle_state"), raw=text,
        )

    def _llm_disambiguate(self, cmd: Command, snap: Snapshot,
                          match: MatchResult) -> Optional[UIElement]:
        try:
            listing = "\n".join(el.brief() for el in match.candidates)
            data = self.llm.chat_json(
                system=(
                    "Pick the UI element a test step refers to. Reply with its "
                    "index, or null if none of them match."
                ),
                user=f"Step: {cmd.raw}\nCandidate elements:\n{listing}",
                schema=DISAMBIG_SCHEMA,
            )
        except LLMUnavailable:
            return None
        idx = data.get("element_index")
        if idx is None:
            return None
        return next((e for e in match.candidates if e.index == idx), None)

    async def _vision_fallback(self, cmd: Command, result: StepResult) -> V.Verification:
        """Last resort: appium-mcp's own vision tool acts from a screenshot."""
        result.llm_used = True
        try:
            await self.mcp.call("ai", {"instruction": cmd.raw})
            await self.dev.stable_snapshot()
            return V.Verification(True, "performed via vision fallback", weak=True)
        except ToolCallError as exc:
            return V.Verification(False, f"vision fallback failed: {exc}")


# ---------------------------------------------------------------------------
def load_flow(path: str) -> List[str]:
    p = Path(path)
    text = p.read_text()
    if p.suffix in (".yaml", ".yml"):
        import yaml
        data = yaml.safe_load(text)
        if isinstance(data, dict):
            data = data.get("steps", [])
        return [str(s) for s in data]
    return [line for line in text.splitlines() if line.strip()]


async def run(cfg: Config, flow_path: str) -> RunReport:
    agent = MobileTestAgent(cfg)
    await agent.start()
    try:
        report = await agent.run_flow(load_flow(flow_path))
    finally:
        await agent.close()
    if cfg.report_file:
        Path(cfg.report_file).write_text(json.dumps(report.as_dict(), indent=2))
    return report

File: mobile-test-agent/src/agent.ts
/**
 * Orchestrator: the find → act → verify loop.
 *
 * Escalation ladder per step (cheapest first — this is the token budget):
 *
 *   parse:   grammar (0 tokens) ────────────► LLM parse (~150 tokens, rare)
 *   find:    fuzzy matcher (0 tokens) ─► scroll-to-find (0) ─► LLM pick
 *            (~300 tokens, rare) ─► appium_ai vision (optional)
 *   act:     deterministic executors (0 tokens)
 *   verify:  postcondition on fresh snapshot (0 tokens)
 *   recover: wait-for-idle → re-find → alternate action path → retry (×N)
 *
 * A typical flow therefore costs *zero* LLM tokens; the LLM only pays for
 * genuinely ambiguous screens, which is how this stays far below
 * plan-every-step agents (DroidRun averages ~3,225 tokens per task).
 */
import { readFileSync, writeFileSync } from "node:fs";
import YAML from "yaml";
import {
  ActionError, DeviceFacade, scrollToFind, selectOption, setDate,
  setToggle, sleep,
} from "./actions.js";
import { loadCapabilities, type Config } from "./config.js";
import {
  COMMAND_SCHEMA, DISAMBIG_SCHEMA, LLMUnavailableError, OllamaLLM,
  type TokenUsage,
} from "./llm.js";
import { confident, resolveTarget, type MatchResult } from "./matcher.js";
import { AppiumMcp, ToolCallError } from "./mcpClient.js";
import {
  parseDateValue, parseStep, MONTH_NAMES,
  type Command, type CommandKind,
} from "./parser.js";
import { brief, scrollContainers, type Snapshot, type UIElement } from "./snapshot.js";
import * as V from "./verify.js";

export type StepStatus = "PASSED" | "PASSED_UNVERIFIED" | "FAILED" | "SKIPPED";

export interface StepResult {
  step: string;
  status: StepStatus;
  attempts: number;
  detail: string;
  durationMs: number;
  llmUsed: boolean;
}

export interface RunReport {
  passed: boolean;
  steps: StepResult[];
  tokenUsage: TokenUsage;
}

export class MobileTestAgent {
  readonly mcp: AppiumMcp;
  readonly llm: OllamaLLM;
  readonly dev: DeviceFacade;

  constructor(private cfg: Config) {
    this.mcp = new AppiumMcp(cfg);
    this.llm = new OllamaLLM(cfg);
    this.dev = new DeviceFacade(this.mcp, cfg);
  }

  // ---- lifecycle ------------------------------------------------------
  async start(): Promise<void> {
    await this.mcp.start();
    if (this.cfg.createSession) await this.ensureSession();
  }

  private async ensureSession(): Promise<void> {
    const capabilities = loadCapabilities(this.cfg);
    try {
      await this.mcp.tryCall("session", [
        { action: "create", capabilities },
        { action: "create" },
        { capabilities },
      ]);
    } catch (err) {
      console.warn(
        `session create failed (${err}); assuming server auto-manages the session`,
      );
    }
  }

  async close(): Promise<void> {
    await this.mcp.close();
  }

  // ---- flow running -----------------------------------------------------
  async runFlow(steps: string[]): Promise<RunReport> {
    const results: StepResult[] = [];
    for (const stepRaw of steps) {
      const step = stepRaw.trim();
      if (!step || step.startsWith("#")) continue;
      const result = await this.runStep(step);
      results.push(result);
      console.log(`${result.status.padEnd(18)} ${step}  (${result.detail})`);
      if (result.status === "FAILED" && !this.cfg.continueOnFailure) break;
    }
    return {
      passed: results.every((r) => r.status !== "FAILED"),
      steps: results,
      tokenUsage: this.llm.usage,
    };
  }

  async runStep(text: string): Promise<StepResult> {
    const t0 = Date.now();
    const result: StepResult = {
      step: text, status: "FAILED", attempts: 0, detail: "",
      durationMs: 0, llmUsed: false,
    };

    let command = parseStep(text, this.cfg.dayFirstDates);
    if (!command) {
      command = await this.llmParse(text);
      result.llmUsed = !!command;
    }
    if (!command) {
      result.detail = "could not understand step (grammar + LLM failed)";
      result.durationMs = Date.now() - t0;
      return result;
    }

    const deadline = t0 + this.cfg.stepTimeoutMs;
    let lastDetail = "";
    for (let attempt = 1; attempt <= this.cfg.maxAttempts; attempt++) {
      result.attempts = attempt;
      try {
        const verification = await this.attempt(command, result);
        if (verification.ok) {
          result.status = verification.weak ? "PASSED_UNVERIFIED" : "PASSED";
          result.detail = verification.detail;
          result.durationMs = Date.now() - t0;
          return result;
        }
        lastDetail = verification.detail;
      } catch (err) {
        if (!(err instanceof ActionError) && !(err instanceof ToolCallError)) throw err;
        lastDetail = String(err instanceof Error ? err.message : err);
      }
      if (Date.now() > deadline) {
        lastDetail += " (step timeout)";
        break;
      }
      // recovery between attempts: settle UI, drop keyboard, re-snapshot
      await this.dev.hideKeyboard();
      await this.dev.stableSnapshot(3000);
      console.log(`retrying step (attempt ${attempt + 1}): ${text} — ${lastDetail}`);
    }

    result.detail = lastDetail;
    result.durationMs = Date.now() - t0;
    return result;
  }

  // ---- one find→act→verify attempt --------------------------------------
  private async attempt(command: Command, result: StepResult): Promise<V.Verification> {
    const { dev, cfg } = { dev: this.dev, cfg: this.cfg };

    // commands with no target element
    switch (command.kind) {
      case "wait":
        await sleep(parseFloat(command.value ?? "1") * 1000);
        return { ok: true, detail: "waited" };
      case "back":
        await this.mcp.tryCall("device", [
          { action: "back" }, { action: "press_back" }, { key: "back" }]);
        return { ok: true, detail: "back pressed", weak: true };
      case "home":
        await this.mcp.tryCall("device", [{ action: "home" }, { key: "home" }]);
        return { ok: true, detail: "home pressed", weak: true };
      case "hide_keyboard":
        await dev.hideKeyboard();
        return { ok: true, detail: "keyboard hidden", weak: true };
      case "launch":
        await this.mcp.tryCall("app_lifecycle", [
          { action: "launch", appId: command.target },
          { action: "activate", appId: command.target },
          { appId: command.target },
        ]);
        await dev.stableSnapshot();
        return { ok: true, detail: `launched ${command.target}`, weak: true };
      case "press_key":
        await this.mcp.tryCall("keyboard", [
          { action: "press", key: command.value }, { key: command.value }]);
        return { ok: true, detail: `pressed ${command.value}`, weak: true };
      default:
        break;
    }

    // assertions: wait-for-condition rather than instant check
    if (command.kind === "assert_visible" || command.kind === "assert_not_visible") {
      const deadline = Date.now() + cfg.waitForElementTimeoutMs;
      for (;;) {
        const snap = await dev.snapshot();
        const ver = V.verifyVisible(snap, command, cfg.matchAcceptScore);
        if (ver.ok || Date.now() > deadline) return ver;
        await sleep(cfg.stablePollIntervalMs);
      }
    }

    // bare directional scroll
    if (command.kind === "scroll" && !command.target) {
      const snap = await dev.stableSnapshot(3000);
      const containers = scrollContainers(snap);
      await dev.swipeDir(command.direction ?? "down", containers[0], snap);
      return { ok: true, detail: `scrolled ${command.direction}` };
    }

    // ---------- FIND ----------
    let before = await dev.stableSnapshot();
    let match = resolveTarget(command, before, cfg.matchAcceptScore,
      cfg.matchAcceptMargin, cfg.matchStrongScore);
    if (!confident(match) && command.kind !== "select") {
      [match, before] = await scrollToFind(dev, command, cfg);
    }
    if (!confident(match) && match.candidates.length) {
      const picked = await this.llmDisambiguate(command, match);
      if (picked) {
        match = { element: picked, score: 1, margin: 1, candidates: match.candidates };
        result.llmUsed = true;
      }
    }
    if (!confident(match)) {
      if (command.kind === "select" && !command.target) {
        // select can work from the option value alone
      } else if (cfg.enableVisionFallback) {
        return this.visionFallback(command, result);
      } else {
        return {
          ok: false,
          detail: `element not found for '${command.target}' ` +
                  `(best score ${match.score.toFixed(2)})`,
        };
      }
    }
    const el = match.element as UIElement;

    // ---------- ACT + VERIFY ----------
    if (command.kind === "tap" || command.kind === "long_press") {
      if (command.kind === "tap") await dev.tap(el);
      else await dev.longPress(el);
      const after = await dev.stableSnapshot();
      return V.verifyTap(before, after, el);
    }

    if (command.kind === "input" && !command.maybeDate) {
      await dev.typeText(el, command.value ?? "");
      const after = await dev.stableSnapshot(4000);
      const ver = V.verifyInput(after, el, command);
      if (ver.ok) await dev.hideKeyboard();
      return ver;
    }

    if (command.kind === "set_date" || (command.kind === "input" && command.maybeDate)) {
      const ok = await setDate(dev, command, cfg, before, el);
      if (!ok) return { ok: false, detail: "date picker interaction failed" };
      const after = await dev.stableSnapshot(4000);
      const d = parseDateValue(command.value ?? "", cfg.dayFirstDates);
      const variants: string[] = [];
      if (d) {
        const dd = String(d.day).padStart(2, "0");
        const mm = String(d.month).padStart(2, "0");
        const mon = MONTH_NAMES[d.month - 1];
        variants.push(
          `${dd}/${mm}/${d.year}`, `${mm}/${dd}/${d.year}`,
          `${dd} ${mon.slice(0, 3)} ${d.year}`, `${mon} ${d.day}, ${d.year}`,
          `${d.year}-${mm}-${dd}`, `${dd}-${mm}-${d.year}`,
        );
      }
      return V.verifyDateShown(after, el, variants);
    }

    if (command.kind === "toggle") {
      const ok = await setToggle(dev, el, command.toggleState);
      return {
        ok,
        detail: ok ? "toggle state verified" : "toggle did not reach desired state",
      };
    }

    if (command.kind === "select") {
      const ok = await selectOption(dev, command, cfg, match.element);
      if (!ok) return { ok: false, detail: `option '${command.value}' not found` };
      const after = await dev.stableSnapshot(4000);
      const probe: Command = {
        kind: "assert_visible", target: command.value,
        secure: false, maybeDate: false, raw: command.raw,
      };
      const shown = V.verifyVisible(after, probe, cfg.matchAcceptScore);
      return { ok: true, detail: `selected '${command.value}'`, weak: !shown.ok };
    }

    if (command.kind === "scroll") {
      // scroll to <target> — finding it IS the pass
      return { ok: true, detail: `scrolled to '${command.target}'` };
    }

    return { ok: false, detail: `unsupported command kind: ${command.kind}` };
  }

  // ---- LLM escalations ------------------------------------------------
  private async llmParse(text: string): Promise<Command | undefined> {
    let data: {
      kind?: CommandKind;
      target?: string | null;
      value?: string | null;
      direction?: Command["direction"] | null;
      toggle_state?: "on" | "off" | null;
    };
    try {
      data = await this.llm.chatJson(
        "You compile one mobile-UI test step into one JSON command. " +
        "kinds: tap, long_press, input, set_date, select, toggle, scroll, " +
        "assert_visible, assert_not_visible, wait, back, home, " +
        "hide_keyboard, launch, press_key. " +
        "target = element description; value = text/option/date.",
        text,
        COMMAND_SCHEMA,
      );
    } catch (err) {
      if (!(err instanceof LLMUnavailableError)) throw err;
      console.warn(`LLM parse unavailable: ${err.message}`);
      return undefined;
    }
    if (!data.kind) return undefined;
    return {
      kind: data.kind,
      target: data.target ?? undefined,
      value: data.value ?? undefined,
      direction: data.direction ?? undefined,
      toggleState: data.toggle_state ?? undefined,
      secure: false,
      maybeDate: false,
      raw: text,
    };
  }

  private async llmDisambiguate(
    command: Command, match: MatchResult,
  ): Promise<UIElement | undefined> {
    let data: { element_index: number | null };
    try {
      const table = match.candidates.map(brief).join("\n");
      data = await this.llm.chatJson(
        "Pick the UI element a test step refers to. Reply with its index, " +
        "or null if none of them match.",
        `Step: ${command.raw}\nCandidate elements:\n${table}`,
        DISAMBIG_SCHEMA,
      );
    } catch (err) {
      if (!(err instanceof LLMUnavailableError)) throw err;
      return undefined;
    }
    if (data.element_index === null || data.element_index === undefined) return undefined;
    return match.candidates.find((e) => e.index === data.element_index);
  }

  /** Last resort: appium-mcp's own vision tool acts from a screenshot. */
  private async visionFallback(
    command: Command, result: StepResult,
  ): Promise<V.Verification> {
    result.llmUsed = true;
    try {
      await this.mcp.call("ai", { instruction: command.raw });
      await this.dev.stableSnapshot();
      return { ok: true, detail: "performed via vision fallback", weak: true };
    } catch (err) {
      if (!(err instanceof ToolCallError)) throw err;
      return { ok: false, detail: `vision fallback failed: ${err.message}` };
    }
  }
}

// ---------------------------------------------------------------------------
export function loadFlow(path: string): string[] {
  const text = readFileSync(path, "utf8");
  if (/\.ya?ml$/.test(path)) {
    let data = YAML.parse(text);
    if (data && typeof data === "object" && !Array.isArray(data)) {
      data = data.steps ?? [];
    }
    return (data as unknown[]).map(String);
  }
  return text.split("\n").filter((line) => line.trim());
}

export async function run(cfg: Config, flowPath: string): Promise<RunReport> {
  const agent = new MobileTestAgent(cfg);
  await agent.start();
  let report: RunReport;
  try {
    report = await agent.runFlow(loadFlow(flowPath));
  } finally {
    await agent.close();
  }
  if (cfg.reportFile) {
    writeFileSync(cfg.reportFile, JSON.stringify(report, null, 2));
  }
  return report;
}
